### PR TITLE
Adds Fourier coefficient visualization tools to fourier module

### PR DESCRIPTION
**Pull request by glassnotes**
**Context:** This module provides the visualization tools for #1160.

**Description of the Change:** TODO

**Benefits:** TODO

**Possible Drawbacks:** Requires matplotlib; TODO

**Related GitHub Issues:** N/A

### CODE REVIEW

File: doc/code/qml_fourier.rst [hunk 1]
```diff
@@ -123,7 +123,227 @@ for multiple dimensions.
...
+functions given a set of Fourier coefficients:
+:func:`~.pennylane.fourier.fourier_reconstruct_function_1D_plot` and
+:func:`~.pennylane.fourier.fourier_reconstruct_function_2D_plot`. For example,
+``circuit_with_weights`` has two input parameters with respect to which we computed
+Fourier coefficients, so we can plot its output:
+
+.. code::
+
+   fourier_reconstruct_function_2D_plot(coeffs)
+   plt.title("Expectation value for circuit with weights", fontsize=14)
```
Comment (mariaschuld) [inline doc/code/qml_fourier.rst:242]: I would add `x[0]` and `x[1]` here to the axes, so it is immediately clear what the numbers mean

File: doc/code/qml_fourier.rst [hunk 2]
```diff
@@ -123,7 +123,227 @@ for multiple dimensions.
...
+
+
+Visualizing multiple sets of coefficients
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+
+Suppose we have not just one set of coefficients, but many sets due to
+differences in the value of ``weights`` in the circuit above. While the
+underlying frequency spectrum will not change, the values of the coefficients
+themselves will vary. The following code generates multiple sets of coefficients:
```
Comment (mariaschuld) [inline doc/code/qml_fourier.rst:257]:
```suggestion
Suppose we do not want to visualize the Fourier coefficients for a fixed 
`weights` argument in `circuit_with_weights`, but the distribution over sets of Fourier coefficients 
when the weights are randomly sampled. For each `weights` sample we get a different set of coefficients:
```

File: doc/code/qml_fourier.rst [hunk 3]
```diff
@@ -123,7 +123,227 @@ for multiple dimensions.
...
+
+   coeffs = fourier_coefficients(simple_circuit, 1, 2)
+
+   >>> coeffs
+   [0.5 +0.j 0.  +0.j 0.25+0.j 0.25+0.j 0.  +0.j]
+
+   
+   # Set up the axes
+   fig, ax = plt.subplots(2, 1, sharex=True, sharey=True)
+   fourier_bar_plot(coeffs, 1, ax)
```
[the remark this answers is not in the document]
Comment (mariaschuld) [reply doc/code/qml_fourier.rst:179]: Or, better, maybe just `bar_plot`?

File: doc/code/qml_fourier.rst [hunk 4]
```diff
@@ -123,7 +123,227 @@ for multiple dimensions.
...
+    :align: left
+    :width: 100%
+    :target: javascript:void(0);
+
+|
+	     
+A similar option is the :func:`~.pennylane.fourier.fourier_box_plot`, which
+produces a plot of the same format but using a box plot.
+	    
+A vastly different view can obtained using the
```
Comment (mariaschuld) [inline doc/code/qml_fourier.rst:290]:
```suggestion
A different view can obtained using the	     
```

File: doc/code/qml_fourier.rst [hunk 5]
```diff
@@ -123,7 +123,227 @@ for multiple dimensions.
...
+   
+.. image:: ../_static/fourier_vis_radial_box.png
+    :align: center
+    :width: 700px
+    :target: javascript:void(0);
+
+|
+
+The left plot displays the real portion, and the right the imaginary
+portion. The labels on the "spokes" of the wheels represent the particular
```
Comment (mariaschuld) [inline doc/code/qml_fourier.rst:316]:
```suggestion
The left plot displays the real part, and the right the imaginary
part of the distribution over a parametrized quantum circuit's 
Fourier coefficients. The labels on the "spokes" of the wheels represent the particular
```

File: doc/code/qml_fourier.rst [hunk 6]
```diff
@@ -123,7 +123,227 @@ for multiple dimensions.
...
+negative frequencies extend counterclockwise from that point, and the positive
+frequencies increase in the clockwise direction. Such plots allow for a more
+compact representation of a large number of frequencies than the linear violin
+and box plots discussed above. For a large number of frequencies, however, it is
+recommended to disable the frequency labelling by setting ``show_freqs=False``,
+and hiding box plot fliers as was done above.
+
+Finally, for the special case of 1- or 2-dimensional functions, we can use the
+:func:`~.pennylane.fourier.fourier_panel_plot` to plot the distributions of the
+individual coefficients on the complex plane.
```
Comment (mariaschuld) [inline doc/code/qml_fourier.rst:328]:
```suggestion
sampled sets of Fourier coefficients on the complex plane.
```

File: pennylane/fourier/__init__.py [hunk 7]
```diff
@@ -11,6 +11,21 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-r"""Tools to compute the Fourier series representation of quantum circuits."""
+r"""Tools to compute and visualise the Fourier series representation of quantum circuits."""
```
Comment (mariaschuld) [inline pennylane/fourier/__init__.py:14]: I wonder if we should delete this so it does not show in the docs? It is kind of confusing when we just introduced the package...

File: pennylane/fourier/utils.py [hunk 8]
```diff
@@ -18,6 +18,32 @@
 import numpy as np
 
 
+def format_nvec(nvec):
+    """Nice strings representing tuples of integers."""
+
+    if isinstance(nvec, int):
+        return str(nvec)
+
```
Comment (mariaschuld) [inline pennylane/fourier/utils.py:26]: Reminder to test these...

File: pennylane/fourier/visualization.py [hunk 9]
```diff
@@ -0,0 +1,606 @@
...
+        ax[1].tick_params(axis="x", colors="white")  # hack to get rid of ticks but keep grid
+
+    return ax
+
+
+def fourier_box_plot(coeffs, n_inputs, ax, colour_dict=None, show_freqs=True, show_fliers=True):
+    """Plots a set of Fourier coefficients as a box plot.
+
+    Args:
+        coeffs (array[complex]): A single set of Fourier coefficients.
```
Comment (mariaschuld) [inline pennylane/fourier/visualization.py:113]: By the way, I always stumble over this...when I feed samples, I feel the most natural data structure is a list of samples. Do you think we could accept a list here and just parse inside the function?

File: pennylane/fourier/visualization.py [hunk 10]
```diff
@@ -0,0 +1,606 @@
...
+
+    Produces either 2-panel plot in which the left panel represents the real
+    parts of Fourier coefficients, and the right the imaginary parts, or a
+    1-panel plot in which half the panel shows the real part and the other half
+    shows the complex. This method accepts multiple sets of coefficients, and
+    plots the distribution of each coefficient as a boxplot.
+
+    Args:
+        coeffs (np.ndarray): A set of Fourier coefficients. Assumed to be
+            from a full fft transform.
```
Comment (mariaschuld) [inline pennylane/fourier/visualization.py:501]: In scipy the docstrings often give a hint that this is the output of function XYZ. Shall we do that here? can be very useful, especially if we do not tell people what shape to put in!

**Review by mariaschuld — APPROVED** [review]
Very nice, I think the plots look great! I had quite a few comments to rephrase, but I think it's getting there. Does not have to be perfect, since we can review the main branch together in the end!

Shall we exclude the visualisation functions from test coverage somehow? Or shall we just run them and check the correct matplotlib object is returned? Maybe we can count axes or something!

File: doc/code/qml_fourier.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (glassnotes) [reply doc/code/qml_fourier.rst:179]: I wanted to give them names that indicate the module but I suppose `fourier.fourier_bar_plot` looks a little redundant now that I look at it with the module name in front. I like `fourier.coefficients_bar_plot`, `fourier.coefficients_radial_box_plot`, etc.

File: pennylane/fourier/visualization.py (hunk 9 above)
Comment (glassnotes) [reply pennylane/fourier/visualization.py:113]: Ah! This is just a copy-paste error in the docs, thanks for catching that. This function does work with a list of samples. (It also accepts a single sample which is then processed and turned into a list down in line 146.) (The only function that explicitly takes just a single set of samples is `fourier_bar_plot`)

**Pull request merged**